src/lib/snapper/snapshot.rb

```diff
@@ -0,0 +1,199 @@
+
+module Yast
+  ##
+  # This class represents the base class for Snapshots mostly based on Snapper
+  # ones.
+  class Snapshot
+    include Yast::Logger
+
+    attr_reader :number, :pre_number, :timestamp, :user
+
+    attr_accessor :description, :cleanup, :user_data, :config, :communication
+
+    class << self
+      # @return [Hash] Mapping string with available subclass
+      def types
+        { "single" => SingleSnapshot, "post" => PostSnapshot, "pre" => PreSnapshot }
+      end
+
+      # @return default strategy responsable of talk with the underlying system
+      def default_communication(context = nil)
+        SnapshotDBus.new(context)
+      end
+
+      # @return [Array] of Snapshots
+      def all(configs = nil, communication = default_communication)
+        communication.all(configs).map do |attrs|
+          new_by_type(attrs)
+        end
+      end
+
+      def find(number, configs = nil, communication = default_communication)
+        all(configs, communication).find { |s| s.number == number }
+      end
+
+      def new_by_type(attrs)
+        types[attrs[:type]].new(attrs)
+      end
+
+      def get_modified_files(from, to, communication = default_communication)
+        communication.get_modified_files(from, to)
+      end
+
+      def list_configs(communication = default_communication)
+        communication.list_configs
+      end
+
+      def get_config(config, communication = default_communication)
+        communication.get_config(config)
+      end
+    end
+
+    # Generic constructor
+    def initialize(attrs = {})
+      raise "No instantiable class" if self.class == Yast::Snapshot
+      @number = attrs[:number]
+      @timestamp = attrs[:timestamp]
+      @user = attrs[:uid]
+      # CleanUp algorithm
+      @cleanup = attrs[:cleanup] || "timeline"
+      @description = attrs[:description] || ""
+      @user_data = attrs[:user_data] || {}
+      @communication =
+        attrs[:communication] || Snapshot.default_communication(self)
+      @config = attrs[:config] || Snapshot.list_configs
+    end
+
+    def save
+      if number.nil?
+        create
+      else
+        update
+      end
+    end
+
+    def name
+      "Snapshot"
+    end
+
+    def valid?
+      true
+    end
+
+    def pre?
+      false
+    end
+
+    def post?
+      false
+    end
+
+    def single?
+      false
+    end
+
+    def user_data_to_s
+      user_data.map { |k, v| "#{k}=#{v}" }.join(", ")
+    end
+
+    def date
+      timestamp.strftime("%F %T")
+    end
+
+    def update(attrs = {})
+      log.info "Updating #{inspect} with #{attrs.inspect}"
+      attrs.each { |k, v| send("#{k}=", v) }
+
+      communication.update
+
+      true
+    end
+
+    def mount_point
+      communication.mount_point
+    end
+
+    def create(communication = self.communication)
+      num = communication.create
+      num ? Snapshot.find(num, config).number : nil
+    end
+
+    def delete(communication = self.communication)
```

Comment:
In general, will be very strange to call it with a different one than from the selected during object initialization.

The name was derived from one of your first comments. It was called strategy and communication_strategy or something similar was suggested. Currenct strategies for snapshots are FileSystem and Dbus, for me channel is also fine.